docs/source/getting-started/first-feature.rst

```diff
@@ -5,124 +5,178 @@ Your First Feature
 
-Now let's implement our messaging feature in the UI!
+We can now implement our messaging feature in the UI!
 
 Messaging UI
 ============
 
-Our messaging feature has two parts: a form with inputs for selecting friends and composing the message text, and a "feed" of messages that have been sent to you.
-Both parts will be implemented as React components that render on the main screen.
+The UI for messaging will consist of a new *Messages* panel in addition to the *Friends* and *Network* panel.
+This panel will have two parts:
+
+    1. A list of messages you've received with their senders.
+    2. A form with a dropdown menu for friend selection and a text field for composing the message.
+
+We will implement each part as a React component, which we'll name ``MessageList`` and ``MessageEdit`` respectively.
+Let's start with the simpler ``MessageList``.
 
 MessageList Component
 ---------------------
 
-The MessageList component is fairly straight-forward: it queries all ``Message`` contracts and displays their contents as a list.
-Here is the code for the entire component.
+The goal of the ``MessageList`` component is to query all ``Message`` contracts where the ``receiver`` is the current user, and display their contents and senders in a list.
+The entire component is shown below.
+You should copy this into a new ``MessageList.tsx`` file in ``ui/src/components``.
+
+.. TODO Include file in template with placeholder for component logic.
 
 .. literalinclude:: code/ui-after/MessageList.tsx
 
-The ``messagesResult`` tracks the state of ``Message`` contracts on the ledger, where we specify no restrictions on the query.
-We extract the actual message data in ``messages``.
-Note that for any particular user, the ``Message`` query yields exactly the messages that have been either written by or sent to that user.
-This is due to how we modelled the signatories and observers in the ``Message`` template, and means we do not risk a privacy breach coming from the application code.
+First we get the ``username`` of the current user with the ``useParty`` hook.
+Then ``messagesResult`` gets the stream of all ``Message`` contracts where the ``receiver`` is our ``username``.
+The streaming aspect means that we don't need to reload the page when new messages come in.
+We extract the *payload* of every ``Message`` contract (the data as opposed to metadata like the contract ID) in ``messages``.
+The rest of the component simply constructs a React ``List`` element with an item for each message.
+
+There is one important point about privacy here.
+No matter how we write our ``Message`` query in the UI code, it is impossible to break the privacy rules given by the DAML model.
+That is, it is impossible to see a ``Message`` contract of which you are not the ``sender`` or the ``receiver`` (the only parties that can observe the contract).
```

Comment:
This will change in the future when we add explicit disclosure. Let's keep that in mind.